File: templates/home/wget.py
#!/usr/bin/env python3

import getopt
import ssl
import sys
import urllib.request
import ssl

USAGE = """
./wget2.py [options] URL

options:

    -c, --ca-file CA_FILE
        Load a set of CA certificates (bundled as one .pem)file.
        (default: /etc/ssl/certs/ca-certificates.crt)
    
    -h, --help
        Show this help message and exit

    -o, --out-file OUT_FILE
        Save download to OUT_FILE
        (default: print to stdout)
""".strip()

def usage(status):
    sys.stderr.write('%s\n' % USAGE)
    sys.exit(status)


def main(argv):
    shortopts = 'c:ho:'
    longopts = ['ca-file=', 'help', 'out-file=']
    ca_file = '/etc/ssl/certs/ca-certificates.crt'
    out_file = None

    try:
        opts, args = getopt.getopt(argv[1:], shortopts, longopts)
    except getopt.GetoptError as err:
        sys.stderr.write('%s\n' % str(err))
        usage(1)

    for o, a in opts:
        if o in ('-c', '--ca-file'):
            ca_file = a
        elif o in ('-h', '--help'):
            usage(0)
        elif o in ('-o', '--out-file'):
            out_file = a
        else:
            assert False, "unhandled option '%s'" % o

    if len(args) != 1:
        usage(1)

    url = args[0]
    ctx = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
    ctx.load_verify_locations(ca_file)

    if out_file:
        f = open(out_file, 'wb')
    else:
        f = sys.stdout

    resp = urllib.request.urlopen(url, timeout=10, context=ctx)
    data = resp.read(1024)
    while data:
        if out_file:
            f.write(data)
        else:
            f.write(data.decode())
        data = resp.read(1024)

    if out_file:
        f.close()

if __name__ == '__main__':
    main(sys.argv)

File: templates/home/wget.lua
#!/usr/bin/env lua5.3

local io = require 'io'
local os = require 'os'

local M = require 'posix.sys.socket'
local unistd = require 'posix.unistd'

local USAGE = [[
usage: wget.lua URL
]]

local function die(fmt, ...)
    io.stderr:write(fmt:format(...))
    os.exit(1)
end

local function printf(fmt, ...)
    io.stdout:write(fmt:format(...))
end

local r, err = M.getaddrinfo('www.cs.umd.edu', 'http', 
        {family=M.AF_INET, socktype=M.SOCK_STREAM})
if not r then
    error(err)
end

for k,v in pairs(r[1]) do
    print(k,v)
end

local fd = M.socket(M.AF_INET, M.SOCK_STREAM, 0)
local ret, errmsg, errnum = M.connect(fd, r[1])

M.send(fd, 'GET /~smherwig/index.html HTTP/1.0\r\nHost: www.cs.umd.edu\r\n\r\n')
local data = {}
while true do
    local b = M.recv(fd, 1024)
    if not b or #b == 0 then
        break
    end
    io.stdout:write(b)
end

unistd.close(fd)
os.exit(0)
